Src/Plugins/Security/SEC101_017.NpmApiKeyValidator.cs

```diff
@@ -0,0 +1,177 @@
+                                                      ref string failureLevel,
+                                                      ref string fingerprintText,
+                                                      ref string message)
+        {
+            if (!groups.TryGetNonEmptyValue("key", out string key))
+            {
+                return nameof(ValidationState.NoMatch);
+            }
+
+            fingerprintText = new Fingerprint
+            {
+                Key = key,
+            }.ToString();
+
+            return nameof(ValidationState.Unknown);
+        }
+
+        protected override string IsValidDynamicHelper(ref string fingerprintText,
+                                                       ref string message)
+        {
+            var fingerprint = new Fingerprint(fingerprintText);
+
+            string key = fingerprint.Key;
+
+            try
+            {
+                using HttpClient client = CreateHttpClient();
+
+                client.DefaultRequestHeaders.Authorization =
+                    new AuthenticationHeaderValue("Bearer", key);
+
+                using HttpResponseMessage response = client
+                    .GetAsync($"https://registry.npmjs.com/-/npm/v1/tokens", HttpCompletionOption.ResponseHeadersRead)
+                    .GetAwaiter()
+                    .GetResult();
+
+                switch (response.StatusCode)
+                {
+                    case HttpStatusCode.OK:
+                    {
+                        return CheckInformation(response.Content.ReadAsStringAsync().GetAwaiter().GetResult(), key, ref message);
+                    }
+
+                    case HttpStatusCode.Unauthorized:
+                    {
+                        return nameof(ValidationState.Unauthorized);
+                    }
+
+                    default:
+                    {
+                        message += $" An unexpected response code was returned attempting to " +
+                                  $"validate: '{response.StatusCode}'";
+                        break;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                return ReturnUnhandledException(ref message, e);
+            }
+
+            return nameof(ValidationState.Unknown);
+        }
+
+        private static string CheckInformation(string content, string key, ref string message)
+        {
+            TokensRoot tokensRoot = JsonConvert.DeserializeObject<TokensRoot>(content);
+            if (tokensRoot?.Tokens?.Count > 0)
+            {
+                foreach (Object obj in tokensRoot.Tokens)
+                {
+                    if (!key.Contains(obj.Token))
+                    {
+                        continue;
+                    }
+
+                    if (obj.Readonly)
+                    {
+                        message = "The token has read permissions.";
```

Comment:
>token [](start = 39, length = 5)

should we rename the rule to NpmTokenValidator instead of ApiKey?

Reply:
Yes. Also, you might consider using apostrophes for these messages, the token has 'read' permissions. The token has 'automation' permissions. Etc.

---
In reply to: [590425352](https://github.com/microsoft/sarif-pattern-matcher/pull/274#discussion_r590425352) [](ancestors = 590425352)

Reply:
Looks like these are called NPM author tokens? So NpmAuthorToken?

---
In reply to: [590446403](https://github.com/microsoft/sarif-pattern-matcher/pull/274#discussion_r590446403) [](ancestors = 590446403,590425352)